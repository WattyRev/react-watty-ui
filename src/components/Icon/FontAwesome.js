import { library } from "@fortawesome/fontawesome-svg-core";
import {
  faAngleDown,
  faCog,
  faEdit,
  faEllipsisH,
  faEraser,
  faExclamationCircle,
  faEye,
  faEyeSlash,
  faHome,
  faImage,
  faInfo,
  faMagic,
  faPalette,
  faPlus,
  faRedo,
  faSave,
  faShare,
  faSignOutAlt,
  faSpinner,
  faUndo,
  faTimes,
  faTrash
} from "@fortawesome/free-solid-svg-icons";

export function setupFontAwesome() {
  library.add(
    faAngleDown,
    faCog,
    faEdit,
    faEllipsisH,
    faEraser,
    faExclamationCircle,
    faEye,
    faEyeSlash,
    faHome,
    faImage,
    faInfo,
    faMagic,
    faPalette,
    faPlus,
    faRedo,
    faSave,
    faShare,
    faSignOutAlt,
    faSpinner,
    faUndo,
    faTimes,
    faTrash
  );
}

/**
 * Render this component at the top of your app to enable the <Icon> component to work.
 */
const FontAwesome = () => {
  setupFontAwesome();
  return null;
};

/** @component */
export default FontAwesome;
