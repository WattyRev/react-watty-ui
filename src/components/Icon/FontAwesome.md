`<FontAwesome/>` does not render anything.
