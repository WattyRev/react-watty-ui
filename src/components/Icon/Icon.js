import React from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import PropTypes from "prop-types";
import { library } from "@fortawesome/fontawesome-svg-core";
import {
  faAngleDown,
  faCog,
  faEdit,
  faEllipsisH,
  faEraser,
  faExclamationCircle,
  faEye,
  faEyeSlash,
  faHome,
  faImage,
  faInfo,
  faMagic,
  faPalette,
  faPlus,
  faRedo,
  faSave,
  faShare,
  faSignOutAlt,
  faSpinner,
  faUndo,
  faTimes,
  faTrash
} from "@fortawesome/free-solid-svg-icons";

export function setupFontAwesome() {
  library.add(
    faAngleDown,
    faCog,
    faEdit,
    faEllipsisH,
    faEraser,
    faExclamationCircle,
    faEye,
    faEyeSlash,
    faHome,
    faImage,
    faInfo,
    faMagic,
    faPalette,
    faPlus,
    faRedo,
    faSave,
    faShare,
    faSignOutAlt,
    faSpinner,
    faUndo,
    faTimes,
    faTrash
  );
}

export const ICON_NAMES = [
  "angle-down",
  "cog",
  "edit",
  "ellipsis-h",
  "eraser",
  "exclamation-circle",
  "eye",
  "eye-slash",
  "home",
  "image",
  "info",
  "magic",
  "palette",
  "plus",
  "redo",
  "save",
  "share",
  "sign-out-alt",
  "spinner",
  "undo",
  "times",
  "trash"
];

const Icon = props => <FontAwesomeIcon {...props} />;

Icon.propTypes = {
  /** The name of the icon to render */
  icon: PropTypes.oneOf(ICON_NAMES)
};

export default Icon;
