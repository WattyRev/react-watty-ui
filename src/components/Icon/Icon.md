Example:

```js
import { ICON_NAMES, Spacer } from "../../index";

<React.Fragment>
  {ICON_NAMES.map(name => (
    <Spacer as="div" bottom="sm" key={name}>
      {name}: <Icon icon={name} />
    </Spacer>
  ))}
</React.Fragment>;
```
