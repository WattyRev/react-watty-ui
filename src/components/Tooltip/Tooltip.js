import React from "react";
import styled, { keyframes } from "styled-components";
import PropTypes from "prop-types";
import BodyPortal from "../utilities/BodyPortal";
import { Text, Icon } from "../../index";

export const fadeIn = keyframes`
  from {
    opacity: 0;
  }
  to {
    opacity: 1;
  }
`;

export const fadeOut = keyframes`
  from {
    opacity: 1;
  }
  to {
    opacity: 0;
  }
`;

/**
 * A style wrapper that creates the tooltip UI.
 * @param {Boolean} removing If true, the tooltip will begin to fade out
 * @param {Number} top The top position of the target element for this to be positioned near
 * @param {Number} left The left position of the target element for this to be positioned near
 * @param {Number} fadeSpeed The speed at which the tooltip should fade in or out (seconds)
 */
export const StyleWrapper = styled.div`
  position: absolute;
  background: ${props => props.theme.colors.white};
  padding: 4px;
  border: 1px solid ${props => props.theme.colors.silver};
  border-radius: 2px;
  box-shadow: 0px 2px 2px -1px ${props => props.theme.colors.gray};
  top: ${props => props.top}px;
  left: ${props => props.left}px;
  transform: translateY(calc(-100% - 10px)) translateX(-6px);
  opacity: ${props => (props.removing ? 0 : 1)};
  animation: ${props => (props.removing ? fadeOut : fadeIn)}
    ${props => props.fadeSpeed}s;
  z-index: 200;

  &:before {
    content: "";
    display: block;
    width: 12px;
    height: 12px;
    background: ${props => props.theme.colors.white};
    border: 1px solid ${props => props.theme.colors.silver};
    border-left: 0px;
    border-top: 0px;
    position: absolute;
    top: calc(100% - 6px);
    left: 5px;
    transform: rotate(45deg);
    border-bottom-right-radius: 2px;
  }
`;

export const TooltipIcon = styled(Icon)`
  border: 1px solid ${props => props.theme.colors.gray};
  color: gray;
  border-radius: 50%;
  width: 12px !important;
  height: 12px;
  display: inline-block;
  padding: 1px;
  box-sizing: border-box;
  transform: translateY(-3px);
  cursor: help;
`;

const TooltipText = styled(Text)`
  position: relative;
`;

/**
 * A component that displays a small info icon. When hovering over the icon, a tooltip will appear
 * floating above that icon, containing the provided content.
 */
class Tooltip extends React.Component {
  static defaultProps = {
    fadeSpeedMs: 250
  };
  static propTypes = {
    /**
     * The speed of the fade animation in milliseconds.
     */
    fadeSpeedMs: PropTypes.number.isRequired,

    /**
     * The content to be displayed in the tooltip
     */
    children: PropTypes.node.isRequired
  };
  state = {
    /**
     * Indicates if the tooltip should be rendered to the dom.
     * @type {Boolean}
     */
    renderTooltip: false,
    /**
     * Indicates that we are about to remove the tooltip from the dom. Allows for the tooltip to
     * fade out before being removed.
     * @type {Boolean}
     */
    removingTooltip: false,
    /**
     * The top position of the info icon, used for targeting the tooltip.
     */
    top: null,
    /**
     * The left position of the info icon, used for targeting the tooltip.
     */
    left: null
  };

  /**
   * Clean up any pending timeouts when unmounting.
   */
  componentWillUnmount() {
    window.clearTimeout(this.fadeTimer);
  }

  /**
   * The stored fade out timer. Allows for cancellation.
   */
  fadeTimer = null;

  /**
   * Displays the tooltip when the info icon is hovered over.
   * @param  {Object} event The mouseenter event from the DOM
   */
  handleMouseEnter = event => {
    window.clearTimeout(this.fadeTimer);
    const targetBounds = event.target.getBoundingClientRect();
    this.setState({
      renderTooltip: true,
      removingTooltip: false,
      top: targetBounds.y,
      left: targetBounds.x
    });
  };

  /**
   * Removes the tooltip when the info icon is no longer hovered over.
   */
  handleMouseLeave = () => {
    this.setState({
      removingTooltip: true
    });
    this.fadeTimer = window.setTimeout(() => {
      this.setState({
        removingTooltip: false,
        renderTooltip: false
      });
    }, this.props.fadeSpeedMs);
  };

  render() {
    return (
      <React.Fragment>
        <TooltipIcon
          icon="info"
          onMouseEnter={this.handleMouseEnter}
          onMouseLeave={this.handleMouseLeave}
        />
        {this.state.renderTooltip && (
          <BodyPortal>
            <StyleWrapper
              className="testing_tooltip"
              top={this.state.top}
              left={this.state.left}
              isVisible={this.state.displayTooltip}
              fadeSpeed={this.props.fadeSpeedMs / 1000}
              removing={this.state.removingTooltip}
            >
              <TooltipText>{this.props.children}</TooltipText>
            </StyleWrapper>
          </BodyPortal>
        )}
      </React.Fragment>
    );
  }
}

export default Tooltip;
