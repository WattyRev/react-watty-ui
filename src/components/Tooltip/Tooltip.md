Example:

```js
<Tooltip>Useful advice</Tooltip>
```
