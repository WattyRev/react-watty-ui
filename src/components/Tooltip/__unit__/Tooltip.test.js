import React from "react";
import { shallow, mount } from "enzyme";
import Theme from "../../theme";
import { setupFontAwesome } from "../Icon";
import Tooltip, {
  StyleWrapper,
  TooltipIcon,
  fadeIn,
  fadeOut
} from "../Tooltip";

describe("Tooltip", () => {
  beforeEach(() => {
    setupFontAwesome();
  });
  describe("StyleWrapper", () => {
    let props;
    beforeEach(() => {
      props = {
        theme: Theme,
        top: 10,
        left: 10,
        removing: false,
        fadeSpeed: 0.2
      };
    });
    it("renders", () => {
      expect.assertions(1);
      const wrapper = mount(<StyleWrapper {...props} />);
      expect(wrapper.find("div")).toHaveLength(1);
    });
    describe("top", () => {
      it("sets the top styling based on the provided value", () => {
        props.top = 100;
        const wrapper = mount(<StyleWrapper {...props} />);
        expect(wrapper).toHaveStyleRule("top", "100px");
      });
    });
    describe("left", () => {
      it("sets the left styling based on the provided value", () => {
        props.left = 11;
        const wrapper = mount(<StyleWrapper {...props} />);
        expect(wrapper).toHaveStyleRule("left", "11px");
      });
    });
    describe("fadeSpeed", () => {
      it("sets the animation speed based on the provied fadeSpeed value", () => {
        props.fadeSpeed = 0.25;
        const wrapper = mount(<StyleWrapper {...props} />);
        expect(wrapper).toHaveStyleRule(
          "animation",
          expect.stringContaining(".25s")
        );
      });
    });
    describe("removing", () => {
      describe("if true", () => {
        beforeEach(() => {
          props.removing = true;
        });
        it("makes the tooltip transparent", () => {
          const wrapper = mount(<StyleWrapper {...props} />);
          expect(wrapper).toHaveStyleRule("opacity", "0");
        });
        it("makes the tooltip fade out", () => {
          const wrapper = mount(<StyleWrapper {...props} />);
          expect(wrapper).toHaveStyleRule(
            "animation",
            expect.stringContaining(fadeOut.getName())
          );
        });
      });
      describe("if false", () => {
        beforeEach(() => {
          props.removing = false;
        });
        it("makes the tooltip opaque", () => {
          const wrapper = mount(<StyleWrapper {...props} />);
          expect(wrapper).toHaveStyleRule("opacity", "1");
        });
        it("makes the tooltip fade in", () => {
          const wrapper = mount(<StyleWrapper {...props} />);
          expect(wrapper).toHaveStyleRule(
            "animation",
            expect.stringContaining(fadeIn.getName())
          );
        });
      });
    });
  });
  describe("TooltipIcon", () => {
    it("renders", () => {
      expect.assertions(1);
      const wrapper = mount(<TooltipIcon theme={Theme} icon="info" />);
      expect(wrapper.find("svg")).toHaveLength(1);
    });
  });

  it("renders", () => {
    expect.assertions(1);
    const wrapper = shallow(<Tooltip>hello</Tooltip>);
    expect(wrapper).toMatchSnapshot();
  });

  it("displays the tooltip when hovering over the icon", () => {
    expect.assertions(1);
    const wrapper = shallow(<Tooltip>hello</Tooltip>);
    wrapper.find(TooltipIcon).simulate("mouseEnter", {
      target: {
        getBoundingClientRect: jest.fn(() => ({
          x: 10,
          y: 11
        }))
      }
    });
    expect(wrapper.find(".testing_tooltip")).toHaveLength(1);
  });
  it("removes the tooltip when the mouse leaves the icon", () => {
    expect.assertions(3);
    const wrapper = shallow(<Tooltip fadeSpeedMs={1}>hello</Tooltip>);
    const icon = wrapper.find(TooltipIcon);
    icon.simulate("mouseEnter", {
      target: {
        getBoundingClientRect: jest.fn(() => ({
          x: 10,
          y: 11
        }))
      }
    });
    expect(wrapper.find(".testing_tooltip")).toHaveLength(1);
    icon.simulate("mouseLeave", {});
    expect(wrapper.find(".testing_tooltip")).toHaveLength(1);
    return new Promise(resolve => {
      window.setTimeout(() => {
        expect(wrapper.find(".testing_tooltip")).toHaveLength(0);
        resolve();
      }, 10);
    });
  });
});
