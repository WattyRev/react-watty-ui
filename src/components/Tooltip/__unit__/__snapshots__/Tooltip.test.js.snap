// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip renders 1`] = `
<Fragment>
  <Styled(Icon)
    icon="info"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  />
</Fragment>
`;
